### PR TITLE
perf(ingestion): prune inert local value symbols

### PR DESCRIPTION
**Pull request by abhigyanpatwari**
## Summary
- Add a post-scope-resolution pruning pass for inert local value symbols.
- Remove only block-scope `Const`/`Variable`/`Static` nodes whose only incident edge is the structural `File -> DEFINES -> symbol` edge.
- Preserve module/file-scope symbols, function/class-like locals, explicit scope-graph definitions, and any local with semantic edges.
- Run the pruning phase before MRO/community/process extraction, with `GITNEXUS_KEEP_LOCAL_VALUE_SYMBOLS=1` as an escape hatch.
- Follow-up fix: block scope is now authoritative even if parser export metadata marks locals inside exported functions as `isExported: true`.

## Validation
- `npm run build`
- `npx tsc --noEmit`
- `npm test -- test/unit/local-symbol-pruner.test.ts`
- `npm test -- test/integration/local-symbol-pruner-pipeline.test.ts`
- `npm test -- test/integration/go-multi-name-worker-metadata.test.ts`
- `npm test -- test/integration/parse-impl-large-fixture.test.ts`
- `npm test -- test/integration/worker-pool.test.ts`
- Full `npm test` was attempted locally. The first run failed broadly because workers were starting from stale build artifacts; after `npm run build`, worker-path suites above pass. Separate local blockers still reproduce in files not touched by this PR, including `test/unit/scope-resolution/go/go-captures-golden.test.ts` digest drift and `test/integration/literal-collectors.test.ts` expecting a scoped collector entry that is undefined.

## Notes
- GitNexus MCP impact/detect tools were not exposed in this Codex thread, so I could not run the repo-index impact check. I kept the code change scoped to ingestion graph cleanup and covered the pruning boundaries with focused tests.

---

> [!NOTE]
> **Medium Risk**
> Changes indexed graph shape (fewer local symbols) and pipeline ordering for downstream phases; scope logic was adjusted for Python class bodies, which could affect what survives pruning.
> 
> **Overview**
> Adds a **`pruneLocalSymbols`** ingestion phase after scope resolution that removes inert block-local **`Const` / `Variable` / `Static`** nodes that only have a structural **`File → DEFINES`** edge, so MRO, communities, and processes run on a slimmer graph. Pruning can be turned off via **`GITNEXUS_KEEP_LOCAL_VALUE_SYMBOLS`** or **`PipelineOptions.keepLocalValueSymbols`**; it still runs when **`skipGraphPhases`** is set.
> 
> **Scope classification fix:** generic variable extraction no longer treats Python class-body **`block`** nodes as function-local scope, so class attributes are not mis-pruned.
> 
> Docs, golden fixture counts, and unit/integration tests cover prune boundaries and preserved **`CALLS`** resolution.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d616787abb333d60bba22eb29ffb2fb25e80d54d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by abhigyanpatwari** [comment]
@claude review this PR

File: gitnexus/src/core/ingestion/pipeline.ts [hunk 1]
```diff
@@ -141,6 +143,7 @@ function buildPhaseList(options?: PipelineOptions): PipelinePhase[] {
     ormPhase,
     crossFilePhase,
     scopeResolutionPhase,
+    pruneLocalSymbolsPhase,
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/pipeline.ts:159]: **🔴 P1 · blocking — regenerate the golden snapshot**  `[reproduced via CI]`

Inserting `pruneLocalSymbolsPhase` here changes default pipeline output: it removes the 4 inert block-scope `Const` (`validated`, `saved`, `clean`, `entry`) + their `File→DEFINES` edges from the `mini-repo` fixture. The committed golden `test/fixtures/pipeline-golden/mini-repo/expected-graph.json` wasn't regenerated, so `pipeline-graph-golden.test.ts` fails in **both** red CI jobs (`Const 4→0`, `DEFINES 20→16`, `nodeCount 37→33`). This — not the unrelated blockers cited in the PR body — is the actual CI failure. **Fix:** run that test with `UPDATE_GOLDEN=1` and commit the regenerated golden (only those counts + `edgeDigest` should change). Corroborated by Codex (independent engine).

File: gitnexus/src/core/ingestion/local-symbol-pruner.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+const emptyStats = (skippedByEnv: boolean): LocalSymbolPruneStats => ({
+  candidateNodes: 0,
+  prunedNodes: 0,
+  keptWithSemanticEdges: 0,
+  skippedByEnv,
+});
+
+const isLocalValueCandidate = (node: GraphNode): boolean => {
+  if (!LOCAL_VALUE_LABELS.has(node.label)) return false;
+  return node.properties.scope === 'block';
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/local-symbol-pruner.ts:27]: **🔴 P1 · silent data loss — `scope === 'block'` is not a reliable cross-language "function-local" signal**  `[code-read]`

`scope` comes from `variable-extractors/generic.ts → determineScope()`, which returns `'block'` on the **first `block` ancestor** — and in tree-sitter-python a **class body is a `block`**. So an untyped Python class attribute (`class Settings:` / `    MAX_SIZE = 100`) is emitted as a `Variable/Const` with `scope='block'` via `@definition.variable` (no class guard); `parse-worker.ts`'s `needsOwner` excludes `Const/Variable/Static`, so it has only a `File→DEFINES` edge → **this pass deletes it.** Any unreferenced / unresolved-access Python class-level constant silently vanishes from the graph. Ruby / Java / C++ / Kotlin / Go are safe (distinct Property/field paths); Python is the only affected language. **Fix:** classify class-body plain assignments as `@definition.property`, or exclude class-body `block` scopes from the candidate filter; add a Python integration test asserting an unreferenced class constant survives. Independently confirmed by Codex + the correctness / adversarial / testing lanes, and predicted by the compiler-design research lane ("class-body vs function-body confusion").

**Review by magyargergo — COMMENTED** [review]
## 🔍 Tri-review digest — `perf(ingestion): prune inert local value symbols`

**Methods:** GitNexus risk + test/CI lanes, Compound-Engineering personas (correctness, adversarial, performance, maintainability, testing), and a compiler-frontend / AST-design research lane — **all Claude** — plus **Codex** (the one independent engine) and a synthesis-critic gate. Engine breakdown: **6 Claude lanes (5 review personas + 1 research) + Codex**. 6 of 7 lanes share a model family, so treat Claude-only agreement as *consistent* and **Codex + Claude agreement as strong**. Codex independently corroborated **both** headline findings. Automated multi-tool digest — verify before acting.

**Verdict — the pruning design is sound, but two issues block merge:** (1) CI is red *because of this PR*, and (2) the pass silently deletes a class of Python symbols.

---

### 🔴 P1 · CI is red because of this PR — golden snapshot not regenerated  `[reproduced via CI]`
Both red jobs (`tests / macos-latest (platform-sensitive)`, `tests / ubuntu / coverage`) fail on the **same** test — `test/integration/pipeline-graph-golden.test.ts`:
```
symbols (nodeCount): expected 37, got 33
byType.Const:        expected 4,  got 0
byRelType.DEFINES:   expected 20, got 16
```
Those deltas are exactly the **4 inert block-scope `Const`** this PR prunes from the `mini-repo` fixture (`validated`, `saved` in `handler.ts`; `clean` in `middleware.ts`; `entry` in `logger.ts`) plus their 4 `File→DEFINES` edges. The committed golden `test/fixtures/pipeline-golden/mini-repo/expected-graph.json` was not regenerated. **The PR body misattributes the CI red to unrelated blockers** (`go-captures-golden`, `literal-collectors`) — the actual failure is this golden drift, 100% caused by the prune. The CI diff lists *only* those 4 counts, so `CALLS`/processes/all other types are untouched — the prune is surgical. **Fix:** run the test with `UPDATE_GOLDEN=1` and commit the regenerated golden (confirm only those counts + `edgeDigest` change). All 4 are genuine function-locals, so the regen does **not** paper over the Python issue below.

### 🔴 P1 · Python class-level constants/attributes are silently pruned  `[code-read — Codex (independent) + correctness/adversarial/testing lanes + compiler theory all converged]`
`isLocalValueCandidate` prunes any `Const/Variable/Static` whose `properties.scope === 'block'`. But `scope` is computed by `variable-extractors/generic.ts → determineScope()`, which returns `'block'` on the **first `block` ancestor** while walking up — and in tree-sitter-python a **class body *is* a `block` node**. So an untyped class attribute:
```python
class Settings:
    MAX_SIZE = 100        # → Const/Variable, scope='block'
```
is classified block-scope. Python emits untyped class attributes via `@definition.variable` (no class guard), and `parse-worker.ts`'s `needsOwner` **excludes** `Const/Variable/Static`, so the node gets **no `HAS_PROPERTY` owner edge — only `File→DEFINES`** → the pass deletes it. Any unreferenced (or unresolved-access) Python class-level value symbol silently disappears from the graph (`class HttpStatus: OK = 200`, config/dataclass-style constants). **Ruby (`body_statement`), Java / C++ / Kotlin (Property/field path), Go are all probe-verified safe — Python is the only affected language.** **Fix:** classify class-body plain assignments as `@definition.property`; or have `determineScope` / the candidate filter not treat a class-body `block` as block scope; or guard the prune on a true function/method ancestor. Add a Python integration test (an unreferenced class constant must survive — the mini-repo has none, so regenerating the golden will not catch this).

> **Why the research lane matters here:** the compiler-design review found the approach is *correct, standard practice* — post-resolution dead-symbol elimination (the SSA/DCE rule "a symbol is dead iff its use-set is empty in a **complete** use-def graph"), symbol-table scope teardown, and SCIP's document-local vs. global symbol tiering (Sourcegraph / Glean prune purely-local nodes for the same graph-size reason). The PR satisfies the ordering invariant (prune runs *after* `scopeResolution` / `crossFile`). Tellingly, the **one failure mode the theory flagged — "class-body vs function-body confusion" — is exactly the Python bug above.**

---

### ✅ Credit where due (validated / suspicions refuted)
- **`removeNode` cleans incident edges** — no dangling `File→DEFINES`; the unit test asserts `relationshipCount → 0`.
- **No mutation race** — phases run sequentially (Kahn topological sort, `await` per phase).
- **No premature-prune cascade** — `mro` / `communities` / `processes` only *read* the post-prune graph; they never create edges to `Const/Variable/Static` nodes (adversarial traced this). Pruning before them is safe.
- **Module / file-scope symbols preserved** — the integration test confirms `MODULE_CONST` survives while function-locals are pruned and the resolved `run → send` `CALLS` edge is kept.

### 🟡 Non-blocking (P3)
- **Programmatic escape hatch missing.** `pruneLocalValueSymbols` accepts `{ keepLocalValueSymbols }`, but the phase calls it with no options and `PipelineOptions` has no such field — so long-running hosts (MCP daemon, eval-server) can only toggle via the `GITNEXUS_KEEP_LOCAL_VALUE_SYMBOLS` **env var** (a `process.env` mutation this codebase otherwise avoids for per-call config). Thread it through `PipelineOptions`.
- **`perf` with no measurement.** Adds an `O(N+E)` pass on every analysis — including small repos and `skipGraphPhases:true` runs, where mro/communities/processes (the intended beneficiaries) don't run, so the pass is pure overhead. No before/after numbers. Consider a benchmark and/or skipping the pass under `skipGraphPhases`. (Optional micro-opt: scan `iterRelationshipsByType('DEFINES')` instead of all edges, keeping the candidate-as-source handling.)
- **Docs / clarity.** `ARCHITECTURE.md`'s phase DAG and the `README` env-var table aren't updated for the new phase/flag. `mro.ts` lists `pruneLocalSymbols` as a *data*-dep purely to force ordering (it never reads the output) — a future dev removing it would silently reorder; a first-class `after` field or a comment would be safer. The `isFileDefinesEdgeToCandidate(rel, rel.sourceId)` source-branch call is correct but reads as "exempt the DEFINES edge" when it can never match — a direct "any outgoing edge ⇒ keep" is clearer.

### 🧪 Test gaps
Coverage is TypeScript-only; the integration test uses `skipGraphPhases:true` (never exercises `prune → mro → communities → processes`) and is gated on a built worker. Add: a Python class-constant case, a candidate-as-source case, and the regenerated golden.

### CI
`quality` (typecheck / lint / format), benchmarks, tree-sitter ABI, and packaged-install smoke are green. The two red jobs are the golden drift above.

---
*Independence: 6 of 7 lanes are Claude under different persona prompts (correlated priors); Codex is the only different engine, and it corroborated both P1s. An automated multi-tool digest — please verify before acting.*
